src/vos/vos_dtx.c

```diff
@@ -720,72 +1033,163 @@ vos_dtx_is_normal_entry(struct umem_instance *umm, umem_off_t entry)
 }
 
 static int
-vos_dtx_alloc(struct umem_instance *umm, struct dtx_handle *dth,
-	      umem_off_t rec_umoff, struct vos_dtx_entry_df **dtxp)
+vos_dtx_alloc(struct umem_instance *umm, struct dtx_handle *dth)
 {
-	struct vos_dtx_entry_df	*dtx;
-	struct vos_container	*cont;
-	umem_off_t		 dtx_umoff;
-	d_iov_t		 kiov;
-	d_iov_t		 riov;
-	struct dtx_rec_bundle	 rbund;
-	int			 rc;
+	struct vos_dtx_act_ent		*dae = NULL;
+	struct dtx_batched_cleanup_blob	*bcb = NULL;
+	struct vos_container		*cont;
+	struct vos_cont_df		*cont_df;
+	struct vos_dtx_scm_blob		*dsb;
+	umem_off_t			 dsb_off;
+	d_iov_t				 kiov;
+	d_iov_t				 riov;
+	int				 rc;
 
 	cont = vos_hdl2cont(dth->dth_coh);
 	D_ASSERT(cont != NULL);
 
-	dtx_umoff = umem_zalloc(umm, sizeof(struct vos_dtx_entry_df));
-	if (dtx_is_null(dtx_umoff))
-		return -DER_NOSPACE;
+	cont_df = cont->vc_cont_df;
+	dth->dth_gen = cont->vc_dtx_resync_gen;
+
+	D_ALLOC_PTR(dae);
+	if (dae == NULL)
+		return -DER_NOMEM;
+
+	DAE_XID(dae) = dth->dth_xid;
+	DAE_OID(dae) = dth->dth_oid;
+	DAE_DKEY_HASH(dae) = dth->dth_dkey_hash;
+	DAE_EPOCH(dae) = dth->dth_epoch;
+	DAE_FLAGS(dae) = dth->dth_leader ? DTX_EF_LEADER : 0;
+	DAE_INTENT(dae) = dth->dth_intent;
+	DAE_SRV_GEN(dae) = dth->dth_gen;
+	DAE_LAYOUT_GEN(dae) = dth->dth_gen;
+
+	if (dtx_is_null(cont_df->cd_dtx_active_tail)) {
+		rc = vos_dtx_create_act_table(cont);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	}
+
+	dsb = umem_off2ptr(umm, cont_df->cd_dtx_active_tail);
+	bcb = d_list_entry(cont->vc_batched_cleanup_list.prev,
+			   struct dtx_batched_cleanup_blob, bcb_cont_link);
+	if (dsb->dsb_index >= dsb->dsb_cap) {
+		struct vos_dtx_scm_blob		*dsb_prev = dsb;
+
+		dsb_off = umem_zalloc(umm, DTX_SCM_BLOB_SIZE);
+		if (dtx_is_null(dsb_off)) {
+			D_ERROR("No space to store actvie DTX (2) "DF_DTI"\n",
+				DP_DTI(&DAE_XID(dae)));
+			D_GOTO(out, rc = -DER_NOSPACE);
+		}
+
+		dsb = umem_off2ptr(umm, dsb_off);
+		dsb->dsb_magic = DTX_ACT_BLOB_MAGIC;
+		dsb->dsb_cap = (DTX_SCM_BLOB_SIZE -
+				sizeof(struct vos_dtx_scm_blob)) /
+			       sizeof(struct vos_dtx_act_ent_df);
+
+		bcb = dtx_bcb_alloc(cont, dsb);
```

Comment:
So this is just allocating enough space ?  Why allocate it if you may not need it?   For instance, what happens if the transaction fails such that dsb_off is not allocated anymore?

Reply:
The @bcb and @dsb is 1:1, anytime when we allocate a @dsb, we will allocate a @bcb for it. If we do not allocate the @dsb, we will not allocate @bcb. All the @bcb are linked into a vos_container based list. So even current transaction failed, the new allocated items will NOT leak, it will be released either when the container is closed or related @dsb is freed (batched cleanup).